gbpservice/nfp/configurator/agents/agent_base.py

```diff
@@ -0,0 +1,181 @@
+
+            # Process the first data blob from the service information list.
+            # Get necessary parameters needed for driver method invocation.
+            method = sa_info_list[0]['method']
+            resource = sa_info_list[0]['resource']
+            kwargs = sa_info_list[0]['kwargs']
+            request_info = kwargs['kwargs']['request_info']
+            del kwargs['kwargs']['request_info']
+            context = sa_info_list[0]['context']
+            service_type = kwargs.get('kwargs').get('service_type')
+
+            # Get the service driver and invoke its method
+            driver = self._get_driver(service_type)
+
+            # Service driver should return "success" on successful API
+            # processing. All other return values and exceptions are treated
+            # as failures.
+            result = getattr(driver, method)(context, **kwargs)
+            if result == 'SUCCESS':
+                success = True
+            else:
+                success = False
+        except Exception as err:
+            result = ("Failed to process %s request. %s" %
+                      (method, str(err).capitalize()))
+
+            success = False
+        finally:
+            # Prepare success notification and populate notification data list
+            msg = {
+                'receiver': const.ORCHESTRATOR,
+                'resource': const.ORCHESTRATOR,
+                'method': "network_function_device_notification",
+                'kwargs': [
+                    {
+                        'context': context,
+                        'resource': resource,
+                        'request_info': request_info,
+                        'result': result,
+                    }
+                ]
+            }
+
+            # If the data processed is first one, then prepare notification
+            # dict. Otherwise, append the notification to the kwargs list.
+            # Whether it is a data batch or single data blob request,
+            # notification generated will be single dictionary. In case of
+            # batch, multiple notifications are sent in the kwargs list.
+            if not notification_data:
+                notification_data.update(msg)
+            else:
+                data = {
+                    'context': context,
+                    'resource': resource,
+                    'request_info': request_info,
+                    'result': result
+                }
+                notification_data['kwargs'].append(data)
+
+        if success:
+            # Remove the processed request data blob from the service
+            # information list. APIs will always process first data blob in
+            # the request.
+            sa_info_list.pop(0)
+
+            # Invoke base class method to process further data blobs in the
+            # request
+            self._rpcmgr.forward_request(sa_info_list, notification_data)
```

Comment:
Little confusing to call a method back in RPC handler. You might want want to move the code in forward_request to a method in this class. Please discuss.